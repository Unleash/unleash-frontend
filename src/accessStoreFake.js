import { Map as $MAp } from 'immutable';

export const createFakeStore = (permissions) => {
    return {
        getState: () => ({
            user: 
                new $MAp({
                    permissions
                })
        }),
    }
}