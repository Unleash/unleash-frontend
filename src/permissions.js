export const ADMIN = 'ADMIN';
export const CREATE_FEATURE = 'CREATE_FEATURE';
export const UPDATE_FEATURE = 'UPDATE_FEATURE';
export const DELETE_FEATURE = 'DELETE_FEATURE';
export const CREATE_STRATEGY = 'CREATE_STRATEGY';
export const UPDATE_STRATEGY = 'UPDATE_STRATEGY';
export const DELETE_STRATEGY = 'DELETE_STRATEGY';
export const UPDATE_APPLICATION = 'UPDATE_APPLICATION';
export const CREATE_CONTEXT_FIELD = 'CREATE_CONTEXT_FIELD';
export const UPDATE_CONTEXT_FIELD = 'UPDATE_CONTEXT_FIELD';
export const DELETE_CONTEXT_FIELD = 'DELETE_CONTEXT_FIELD';

export function hasPermission(user, permission) {
    return (
        user &&
        (!user.permissions || user.permissions.indexOf(ADMIN) !== -1 || user.permissions.indexOf(permission) !== -1)
    );
}
