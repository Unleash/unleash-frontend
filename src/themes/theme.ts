import { createTheme } from '@mui/material/styles';
import { colors } from './colors';

export default createTheme({
    boxShadows: {
        main: '0px 2px 4px rgba(129, 122, 254, 0.2)',
    },
    typography: {
        fontFamily: 'Sen, Roboto, sans-serif',
        fontWeightBold: '700',
        fontWeightMedium: '700',
        allVariants: { lineHeight: 1.4 },
        button: { lineHeight: 1.75 },
        h1: {
            fontSize: '1.5rem',
            lineHeight: 1.875,
        },
    },
    fontSizes: {
        mainHeader: '1.25rem',
        subHeader: '1.1rem',
        bodySize: '1rem',
        smallBody: '0.9rem',
        smallerBody: '0.8rem',
    },
    fontWeight: {
        thin: 300,
        medium: 400,
        semi: 700,
        bold: 700,
    },
    shape: {
        borderRadius: '3px',
        borderRadiusLarge: '10px',
        borderRadiusExtraLarge: '25px',
    },
    palette: {
        primary: {
            main: colors.purple[800],
            light: colors.purple[700],
            dark: colors.purple[900],
        },
        secondary: {
            main: colors.purple[800],
            light: colors.purple[700],
            dark: colors.purple[900],
        },
        info: {
            light: colors.blue[700],
            main: colors.blue[700],
            dark: colors.blue[800],
        },
        success: {
            light: colors.green[700],
            main: colors.green[700],
            dark: colors.green[800],
        },
        warning: {
            light: colors.orange[700],
            main: colors.orange[700],
            dark: colors.orange[800],
        },
        error: {
            light: colors.red[700],
            main: colors.red[700],
            dark: colors.red[800],
        },
        divider: colors.grey[300],
        dividerAlternative: colors.grey[500],
        grey: colors.grey,
        text: {
            primary: colors.grey[900],
            secondary: colors.grey[800],
            disabled: colors.grey[600],
        },
        code: {
            main: '#0b8c8f',
            diffAdd: 'green',
            diffSub: 'red',
            diffNeutral: 'black',
            edited: 'blue',
            background: '#efefef',
        },
        activityIndicators: {
            unknown: colors.grey[100],
            recent: colors.green[100],
            inactive: colors.orange[200],
            abandoned: colors.red[200],
        },
    },
    components: {
        MuiTableHead: {
            styleOverrides: {
                root: {
                    background: colors.grey[200],
                },
            },
        },
        MuiLink: {
            styleOverrides: {
                root: {
                    color: colors.purple[900],
                },
            },
        },
        MuiBreadcrumbs: {
            styleOverrides: {
                root: {
                    '& a': {
                        color: colors.purple[900],
                    },
                },
            },
        },
        MuiTableRow: {
            styleOverrides: {
                root: {
                    '&.MuiTableRow-hover:hover': {
                        background: colors.grey[100],
                    },
                },
            },
        },
    },
});
