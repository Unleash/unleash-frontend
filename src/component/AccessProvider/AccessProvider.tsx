import { FC } from "react";

import AccessContext from '../../contexts/AccessContext'
import { ADMIN } from "./permissions";

// TODO: Type up redux store
interface IAccessProvider {
    store: any;
}

interface IPermission {
  permission: string;
  project: string | null;
}

const AccessProvider: FC<IAccessProvider> = ({store, children}) => {
  const hasAccess = (permission: string, project: string) => {
    const permissions = store.getState().user.get('permissions') || [];

    const result = permissions.some((p: IPermission) => {
      if(p.permission === ADMIN) {
        return true
      }
      if(p.permission === permission && p.project === project) {
        return true;
      }
      return false;
    });

    return result;
  };

  const context = { hasAccess };

  return <AccessContext.Provider value={context}>{children}</AccessContext.Provider>
}

export default AccessProvider;