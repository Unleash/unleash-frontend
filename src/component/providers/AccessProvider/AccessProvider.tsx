import { ReactElement, ReactNode, useCallback, useMemo } from 'react';
import AccessContext, { IAccessContext } from '../../../contexts/AccessContext';
import { ADMIN } from './permissions';
import { IPermission } from '../../../interfaces/user';
import { useAuthPermissions } from '../../../hooks/api/getters/useAuth/useAuthPermissions';

interface IAccessProviderProps {
    children: ReactNode;
    permissions: IPermission[];
}

// TODO(olav): Mock useAuth instead of using props.permissions in tests.
const AccessProvider = (props: IAccessProviderProps): ReactElement => {
    const auth = useAuthPermissions();
    const permissions = props.permissions ?? auth.permissions;

    const isAdmin: boolean = useMemo(() => {
        return checkAdmin(permissions);
    }, [permissions]);

    const hasAccess = useCallback(
        (permission: string, project?: string, environment?: string) => {
            return checkPermissions(
                permissions,
                permission,
                project,
                environment
            );
        },
        [permissions]
    );

    const value: IAccessContext = useMemo(
        () => ({
            isAdmin,
            hasAccess,
        }),
        [isAdmin, hasAccess]
    );

    return (
        <AccessContext.Provider value={value}>
            {props.children}
        </AccessContext.Provider>
    );
};

const checkAdmin = (permissions: IPermission[] | undefined): boolean => {
    if (!permissions) {
        return false;
    }

    return permissions.some(p => {
        return p.permission === ADMIN;
    });
};

const checkPermissions = (
    permissions: IPermission[] | undefined,
    permission: string,
    project?: string,
    environment?: string
): boolean => {
    if (!permissions) {
        return false;
    }

    return permissions.some(p => {
        return checkPermission(p, permission, project, environment);
    });
};

const checkPermission = (
    p: IPermission,
    permission: string,
    project?: string,
    environment?: string
): boolean => {
    if (!permission) {
        console.warn(`Missing permission for AccessProvider: ${permission}`)
        return false
    }

    if (p.permission === ADMIN) {
        return true;
    }

    if (
        p.permission === permission &&
        (p.project === project || p.project === '*') &&
        (p.environment === environment || p.environment === '*')
    ) {
        return true;
    }

    if (
        p.permission === permission &&
        (p.project === project || p.project === '*') &&
        p.environment === null
    ) {
        return true;
    }

    return (
        p.permission === permission &&
        p.project === undefined &&
        p.environment === null
    );
};

export default AccessProvider;
