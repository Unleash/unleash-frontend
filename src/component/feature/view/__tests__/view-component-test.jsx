import React from 'react';
import { MemoryRouter } from 'react-router-dom';
import { createStore } from 'redux';

import { Provider } from 'react-redux';

import { ThemeProvider } from '@material-ui/core';
import ViewFeatureToggleComponent from '../../FeatureView/FeatureView';
import renderer from 'react-test-renderer';
import { ADMIN, DELETE_FEATURE, UPDATE_FEATURE } from '../../../AccessProvider/permissions';

import theme from '../../../../themes/main-theme';
import { createFakeStore } from '../../../../accessStoreFake';
import AccessProvider from '../../../AccessProvider/AccessProvider';

jest.mock('../update-strategies-container', () => ({
    __esModule: true,
    default: 'UpdateStrategiesComponent',
}));
jest.mock('../../feature-type-select-container', () => 'FeatureTypeSelect');
jest.mock('../../../common/ProjectSelect', () => 'ProjectSelect');
jest.mock('../../tag-type-select-container', () => 'TagTypeSelect');
jest.mock('../../feature-tag-component', () => 'FeatureTagComponent');
jest.mock('../../add-tag-dialog-container', () => 'AddTagDialog');

test('renders correctly with one feature', () => {
    const feature = {
        name: 'Another',
        description: "another's description",
        enabled: false,
        stale: false,
        type: 'release',
        strategies: [
            {
                name: 'gradualRolloutRandom',
                parameters: {
                    percentage: 50,
                },
            },
        ],
        createdAt: '2018-02-04T20:27:52.127Z',
    };

    const mockStore = {
        projects: {
            toJS: () => [
                { id: 'default', initial: false, name: 'Default' },
                { id: 'myProject', initial: false, name: 'MyProject' },
            ],
        },
        uiConfig: {
            toJS: () => ({
                flags: {
                    P: true,
                },
            }),
        },
    };

    const mockReducer = state => state;

    const tree = renderer.create(
        <MemoryRouter>
            <Provider store={createStore(mockReducer, mockStore)}>
                <ThemeProvider theme={theme}>
                    <AccessProvider store={createFakeStore([{permission: ADMIN}])}>
                    <ViewFeatureToggleComponent
                        activeTab={'strategies'}
                        featureToggleName="another"
                        features={[feature]}
                        featureToggle={feature}
                        fetchFeatureToggles={jest.fn()}
                        history={{}}
                        featureTags={[]}
                        fetchTags={jest.fn()}
                        untagFeature={jest.fn()}
                    />
                    </AccessProvider>
                </ThemeProvider>
            </Provider>
        </MemoryRouter>
    );

    expect(tree).toMatchSnapshot();
});
