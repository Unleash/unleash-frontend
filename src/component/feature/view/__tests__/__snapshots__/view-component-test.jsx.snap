// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders correctly with one feature 1`] = `
<div
  className="MuiPaper-root MuiPaper-elevation1 MuiPaper-rounded"
  style={
    Object {
      "overflow": "visible",
    }
  }
>
  <div>
    <div
      className="header"
    >
      <h1
        className="MuiTypography-root heading MuiTypography-h1"
      >
        Another
      </h1>
      <div
        className="MuiChip-root active"
        onKeyDown={[Function]}
        onKeyUp={[Function]}
        title="Feature toggle is active."
      >
        <span
          className="MuiChip-label"
        >
          Active
        </span>
      </div>
    </div>
    <div
      className="featureInfoContainer makeStyles-contentSpacingY-1"
    >
      <div
        className="MuiFormControl-root"
      >
        <p
          className="MuiTypography-root MuiTypography-body1"
        >
          another's description
          <a
            aria-disabled={false}
            aria-label="toggle description edit"
            className="MuiButtonBase-root MuiIconButton-root"
            href="/#edit"
            onBlur={[Function]}
            onClick={[Function]}
            onDragLeave={[Function]}
            onFocus={[Function]}
            onKeyDown={[Function]}
            onKeyUp={[Function]}
            onMouseDown={[Function]}
            onMouseLeave={[Function]}
            onMouseUp={[Function]}
            onTouchEnd={[Function]}
            onTouchMove={[Function]}
            onTouchStart={[Function]}
            role="button"
            tabIndex={0}
          >
            <span
              className="MuiIconButton-label"
            >
              <svg
                aria-hidden={true}
                className="MuiSvgIcon-root"
                focusable="false"
                viewBox="0 0 24 24"
              >
                <path
                  d="M3 17.25V21h3.75L17.81 9.94l-3.75-3.75L3 17.25zM20.71 7.04c.39-.39.39-1.02 0-1.41l-2.34-2.34a.9959.9959 0 00-1.41 0l-1.83 1.83 3.75 3.75 1.83-1.83z"
                />
              </svg>
            </span>
          </a>
        </p>
      </div>
      <div
        className="selectContainer"
      >
        <FeatureTypeSelect
          label="Feature type"
          onChange={[Function]}
          value="release"
        />
        Â 
        <div
          className="MuiFormControl-root"
        >
          <label
            className="MuiFormLabel-root MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated MuiInputLabel-marginDense MuiInputLabel-outlined"
            data-shrink={false}
          >
            Project
          </label>
          <div
            className="MuiInputBase-root MuiOutlinedInput-root MuiInputBase-formControl MuiInputBase-marginDense MuiOutlinedInput-marginDense"
            onClick={[Function]}
            size="small"
          >
            <div
              aria-haspopup="listbox"
              className="MuiSelect-root MuiSelect-select MuiSelect-selectMenu MuiSelect-outlined MuiInputBase-input MuiOutlinedInput-input MuiInputBase-inputMarginDense MuiOutlinedInput-inputMarginDense"
              onBlur={[Function]}
              onFocus={[Function]}
              onKeyDown={[Function]}
              onMouseDown={[Function]}
              role="button"
              tabIndex={0}
            >
              <span
                dangerouslySetInnerHTML={
                  Object {
                    "__html": "&#8203;",
                  }
                }
              />
            </div>
            <input
              aria-hidden={true}
              className="MuiSelect-nativeInput"
              onAnimationStart={[Function]}
              onChange={[Function]}
              required={false}
              tabIndex={-1}
            />
            <svg
              aria-hidden={true}
              className="MuiSvgIcon-root MuiSelect-icon MuiSelect-iconOutlined"
              focusable="false"
              viewBox="0 0 24 24"
            >
              <path
                d="M7 10l5 5 5-5z"
              />
            </svg>
            <fieldset
              aria-hidden={true}
              className="PrivateNotchedOutline-root-9 MuiOutlinedInput-notchedOutline"
            >
              <legend
                className="PrivateNotchedOutline-legendLabelled-11"
              >
                <span>
                  Project
                </span>
              </legend>
            </fieldset>
          </div>
        </div>
      </div>
      <FeatureTagComponent
        featureToggleName="Another"
        tags={Array []}
        untagFeature={[MockFunction]}
      />
    </div>
  </div>
  <div
    className="actions"
  >
    <span
      style={
        Object {
          "paddingRight": "24px",
        }
      }
    >
      <span
        className="MuiSwitch-root"
      >
        <span
          aria-disabled={false}
          className="MuiButtonBase-root MuiIconButton-root PrivateSwitchBase-root-13 MuiSwitch-switchBase MuiSwitch-colorSecondary"
          onBlur={[Function]}
          onDragLeave={[Function]}
          onFocus={[Function]}
          onKeyDown={[Function]}
          onKeyUp={[Function]}
          onMouseDown={[Function]}
          onMouseLeave={[Function]}
          onMouseUp={[Function]}
          onTouchEnd={[Function]}
          onTouchMove={[Function]}
          onTouchStart={[Function]}
          tabIndex={null}
        >
          <span
            className="MuiIconButton-label"
          >
            <input
              checked={false}
              className="PrivateSwitchBase-input-16 MuiSwitch-input"
              disabled={false}
              onChange={[Function]}
              type="checkbox"
            />
            <span
              className="MuiSwitch-thumb"
            />
          </span>
        </span>
        <span
          className="MuiSwitch-track"
        />
      </span>
      <span>
        Disabled
      </span>
    </span>
    <div>
      <AddTagDialog
        featureToggleName="Another"
      />
      <button
        aria-controls="feature-stale-dropdown"
        aria-haspopup="true"
        className="MuiButtonBase-root MuiButton-root MuiButton-text dropdownButton"
        disabled={false}
        id="feature-stale-dropdown"
        onBlur={[Function]}
        onClick={[Function]}
        onDragLeave={[Function]}
        onFocus={[Function]}
        onKeyDown={[Function]}
        onKeyUp={[Function]}
        onMouseDown={[Function]}
        onMouseLeave={[Function]}
        onMouseUp={[Function]}
        onTouchEnd={[Function]}
        onTouchMove={[Function]}
        onTouchStart={[Function]}
        style={
          Object {
            "fontWeight": "bold",
          }
        }
        tabIndex={0}
        type="button"
      >
        <span
          className="MuiButton-label"
        >
          ACTIVE
          <span
            className="MuiButton-endIcon MuiButton-iconSizeMedium"
          >
            <span
              aria-hidden={true}
              className="material-icons MuiIcon-root"
            >
              arrow_drop_down
            </span>
          </span>
        </span>
      </button>
      <a
        aria-disabled={false}
        className="MuiButtonBase-root MuiButton-root MuiButton-text"
        href="/features/copy/Another"
        onBlur={[Function]}
        onClick={[Function]}
        onDragLeave={[Function]}
        onFocus={[Function]}
        onKeyDown={[Function]}
        onKeyUp={[Function]}
        onMouseDown={[Function]}
        onMouseLeave={[Function]}
        onMouseUp={[Function]}
        onTouchEnd={[Function]}
        onTouchMove={[Function]}
        onTouchStart={[Function]}
        role="button"
        tabIndex={0}
        title="Create new feature toggle by cloning configuration"
      >
        <span
          className="MuiButton-label"
        >
          Clone
        </span>
      </a>
      <button
        className="MuiButtonBase-root MuiButton-root MuiButton-text"
        disabled={false}
        onBlur={[Function]}
        onClick={[Function]}
        onDragLeave={[Function]}
        onFocus={[Function]}
        onKeyDown={[Function]}
        onKeyUp={[Function]}
        onMouseDown={[Function]}
        onMouseLeave={[Function]}
        onMouseUp={[Function]}
        onTouchEnd={[Function]}
        onTouchMove={[Function]}
        onTouchStart={[Function]}
        style={
          Object {
            "flexShrink": 0,
          }
        }
        tabIndex={0}
        title="Archive feature toggle"
        type="button"
      >
        <span
          className="MuiButton-label"
        >
          Archive
        </span>
      </button>
    </div>
  </div>
  <hr />
  <div
    className="MuiPaper-root makeStyles-tabNav-17 MuiPaper-elevation1 MuiPaper-rounded"
  >
    <div
      className="MuiTabs-root"
    >
      <div
        className="MuiTabs-scroller MuiTabs-fixed"
        onScroll={[Function]}
        style={
          Object {
            "marginBottom": null,
            "overflow": "hidden",
          }
        }
      >
        <div
          className="MuiTabs-flexContainer MuiTabs-centered"
          onKeyDown={[Function]}
          role="tablist"
        >
          <button
            aria-controls="tabpanel-0"
            aria-selected={true}
            className="MuiButtonBase-root MuiTab-root MuiTab-textColorPrimary Mui-selected"
            disabled={false}
            id="tab-0"
            onBlur={[Function]}
            onClick={[Function]}
            onDragLeave={[Function]}
            onFocus={[Function]}
            onKeyDown={[Function]}
            onKeyUp={[Function]}
            onMouseDown={[Function]}
            onMouseLeave={[Function]}
            onMouseUp={[Function]}
            onTouchEnd={[Function]}
            onTouchMove={[Function]}
            onTouchStart={[Function]}
            role="tab"
            tabIndex={0}
            type="button"
          >
            <span
              className="MuiTab-wrapper"
            >
              Activation
            </span>
            <span
              className="PrivateTabIndicator-root-18 PrivateTabIndicator-colorPrimary-19 MuiTabs-indicator"
              style={Object {}}
            />
          </button>
          <button
            aria-controls="tabpanel-1"
            aria-selected={false}
            className="MuiButtonBase-root MuiTab-root MuiTab-textColorPrimary"
            disabled={false}
            id="tab-1"
            onBlur={[Function]}
            onClick={[Function]}
            onDragLeave={[Function]}
            onFocus={[Function]}
            onKeyDown={[Function]}
            onKeyUp={[Function]}
            onMouseDown={[Function]}
            onMouseLeave={[Function]}
            onMouseUp={[Function]}
            onTouchEnd={[Function]}
            onTouchMove={[Function]}
            onTouchStart={[Function]}
            role="tab"
            tabIndex={-1}
            type="button"
          >
            <span
              className="MuiTab-wrapper"
            >
              Metrics
            </span>
          </button>
          <button
            aria-controls="tabpanel-2"
            aria-selected={false}
            className="MuiButtonBase-root MuiTab-root MuiTab-textColorPrimary"
            disabled={false}
            id="tab-2"
            onBlur={[Function]}
            onClick={[Function]}
            onDragLeave={[Function]}
            onFocus={[Function]}
            onKeyDown={[Function]}
            onKeyUp={[Function]}
            onMouseDown={[Function]}
            onMouseLeave={[Function]}
            onMouseUp={[Function]}
            onTouchEnd={[Function]}
            onTouchMove={[Function]}
            onTouchStart={[Function]}
            role="tab"
            tabIndex={-1}
            type="button"
          >
            <span
              className="MuiTab-wrapper"
            >
              Variants
            </span>
          </button>
          <button
            aria-controls="tabpanel-3"
            aria-selected={false}
            className="MuiButtonBase-root MuiTab-root MuiTab-textColorPrimary"
            disabled={false}
            id="tab-3"
            onBlur={[Function]}
            onClick={[Function]}
            onDragLeave={[Function]}
            onFocus={[Function]}
            onKeyDown={[Function]}
            onKeyUp={[Function]}
            onMouseDown={[Function]}
            onMouseLeave={[Function]}
            onMouseUp={[Function]}
            onTouchEnd={[Function]}
            onTouchMove={[Function]}
            onTouchStart={[Function]}
            role="tab"
            tabIndex={-1}
            type="button"
          >
            <span
              className="MuiTab-wrapper"
            >
              Log
            </span>
          </button>
        </div>
      </div>
    </div>
  </div>
  <div
    className="tabContentContainer"
  >
    <div
      aria-labelledby="wrapped-tab-0"
      hidden={false}
      id="wrapped-tabpanel-0"
      role="tabpanel"
    >
      <UpdateStrategiesComponent
        featureToggle={
          Object {
            "createdAt": "2018-02-04T20:27:52.127Z",
            "description": "another's description",
            "enabled": false,
            "name": "Another",
            "stale": false,
            "strategies": Array [
              Object {
                "name": "gradualRolloutRandom",
                "parameters": Object {
                  "percentage": 50,
                },
              },
            ],
            "type": "release",
          }
        }
        features={
          Array [
            Object {
              "createdAt": "2018-02-04T20:27:52.127Z",
              "description": "another's description",
              "enabled": false,
              "name": "Another",
              "stale": false,
              "strategies": Array [
                Object {
                  "name": "gradualRolloutRandom",
                  "parameters": Object {
                    "percentage": 50,
                  },
                },
              ],
              "type": "release",
            },
          ]
        }
        history={Object {}}
      />
    </div>
    <div
      aria-labelledby="wrapped-tab-1"
      hidden={true}
      id="wrapped-tabpanel-1"
      role="tabpanel"
    />
    <div
      aria-labelledby="wrapped-tab-2"
      hidden={true}
      id="wrapped-tabpanel-2"
      role="tabpanel"
    />
    <div
      aria-labelledby="wrapped-tab-3"
      hidden={true}
      id="wrapped-tabpanel-3"
      role="tabpanel"
    />
  </div>
</div>
`;
